indexer/disk-saver/src/persistence.rs

```diff
@@ -0,0 +1,248 @@
+pub struct DiskPersistence {
+    pub file_path: PathBuf,
+    pub compressed: bool,
+}
+
+impl DiskPersistence {
+    /// Will automatically set the file extension based on the serialization and compression
+    pub fn new(mut file_path: PathBuf, mut should_compress: bool) -> Self {
+        // For if I'll support other serialization mechanism so we can figure
+        // out the serialization format based on the file extension.
+        file_path.set_extension("borsh");
+
+        let mut compressed_file_path = file_path.clone();
+        compressed_file_path.set_extension("borsh.xz");
+
+        match (
+            should_compress,
+            file_path.exists(),
+            compressed_file_path.exists(),
+        ) {
+            (_, true, true) => {
+                #[cfg(feature = "tracing")]
+                tracing::error!(
+                    file_path = %file_path.display(),
+                    compressed_file_path = %compressed_file_path.display(),
+                    "Both compressed and uncompressed file exists. Keeping uncompressed."
+                );
+
+                std::fs::remove_file(&compressed_file_path).ok();
+
+                should_compress = false;
+            },
+            // Compressed file exists, we'll use it.
+            (_, _, true) => {
+                should_compress = true;
+                file_path = compressed_file_path;
+            },
+            // Uncompressed file exists, we'll use it.
+            (_, true, _) => {
+                should_compress = false;
+            },
+            // No file exists, we'll create an compressed file.
+            (true, ..) => {
+                file_path = compressed_file_path;
+            },
+            // No file exists, we'll create an uncompressed file.
+            (false, ..) => {},
+        }
+
+        Self {
+            file_path,
+            compressed: should_compress,
+        }
+    }
+
+    /// Serialize and compress the data and save it to disk.
+    pub fn save<T: BorshSerialize>(&self, data: &T) -> Result<(), Error> {
+        // Block are final, we don't want to overwrite them.
+        // NOTE: we might need a `force: true` option in the future but I don't see
+        // why we would want to overwrite a block.
+        if self.file_path.exists() {
+            return Err(Error::AlreadyExists(self.file_path.clone()));
+        }
+
+        let serialized = data.to_borsh_vec()?;
+
+        let parent = Path::new(&self.file_path)
+            .parent()
+            .unwrap_or_else(|| Path::new("."));
+        let mut tmp_file = NamedTempFile::new_in(parent)?;
+
+        if self.compressed {
+            let mut compressed = Vec::new();
+            lzma_compress(&mut serialized.as_slice(), &mut compressed)?;
+            tmp_file.write_all(&compressed)?;
+        } else {
+            tmp_file.write_all(&serialized)?;
+        }
+
+        tmp_file.flush()?;
+        tmp_file.persist(&self.file_path)?;
+        Ok(())
```

Comment:
Add an empty line before `Ok(())`. In general I like to have an empty line before the return value (except for the very small functions)